File: apps/web_companion/src/state/conversation_store.ts
import { createStore, type Store } from './store';
import type { Conversation } from './models';

export interface ConversationState {
  /** Ordered: most recent activity first. */
  conversations: Conversation[];
}

export const createConversationStore = (): Store<ConversationState> =>
  createStore<ConversationState>({ conversations: [] });

/** Insert or update a conversation and re-sort by recent activity. */
export function upsertConversation(
  store: Store<ConversationState>,
  conversation: Conversation,
): void {
  store.set((prev) => {
    const others = prev.conversations.filter((c) => c.key !== conversation.key);
    const existing = prev.conversations.find((c) => c.key === conversation.key);
    const merged: Conversation = { ...existing, ...conversation, unreadLocal: existing ? existing.unreadLocal : conversation.unreadLocal };
    return { conversations: sortConversations([merged, ...others]) };
  });
}

/** Tail update only (no re-insert for unknown conversations). */
export function touchConversation(
  store: Store<ConversationState>,
  key: string,
  preview: string,
  at: number,
): void {
  store.set((prev) => ({
    conversations: sortConversations(
      prev.conversations.map((c) =>
        c.key === key ? { ...c, lastMessagePreview: preview, lastMessageAt: at } : c,
      ),
    ),
  }));
}

export function bumpUnread(store: Store<ConversationState>, key: string): void {
  store.set((prev) => ({
    conversations: prev.conversations.map((c) =>
      c.key === key ? { ...c, unreadLocal: c.unreadLocal + 1 } : c,
    ),
  }));
}

export function clearUnread(store: Store<ConversationState>, key: string): void {
  store.set((prev) => ({
    conversations: prev.conversations.map((c) =>
      c.key === key ? { ...c, unreadLocal: 0 } : c,
    ),
  }));
}

/** Drop a conversation entirely (left the group, or kicked from it). */
export function removeConversation(store: Store<ConversationState>, key: string): void {
  store.set((prev) => {
    if (!prev.conversations.some((c) => c.key === key)) return prev;
    return { conversations: prev.conversations.filter((c) => c.key !== key) };
  });
}

function sortConversations(conversations: Conversation[]): Conversation[] {
  return [...conversations].sort(
    (a, b) => (b.lastMessageAt ?? 0) - (a.lastMessageAt ?? 0) || a.key.localeCompare(b.key),
  );
}

File: apps/mobile_companion/lib/state/conversation_store.dart
import 'models.dart';
import 'store.dart';

class ConversationState {
  const ConversationState({required this.conversations});

  /// Ordered: most recent activity first.
  final List<Conversation> conversations;
}

Store<ConversationState> createConversationStore() =>
    Store<ConversationState>(const ConversationState(conversations: []));

/// Insert or update a conversation and re-sort by recent activity.
void upsertConversation(
    Store<ConversationState> store, Conversation conversation) {
  store.update((prev) {
    final others =
        prev.conversations.where((c) => c.key != conversation.key).toList();
    final existing = prev.conversations
        .where((c) => c.key == conversation.key)
        .toList()
        .firstOrNull;
    final merged = existing == null
        ? conversation
        : Conversation(
            kind: conversation.kind,
            key: conversation.key,
            channelId: conversation.channelId,
            peerId: conversation.peerId,
            title: conversation.title,
            ownerId: conversation.ownerId ?? existing.ownerId,
            unreadLocal: existing.unreadLocal,
            lastMessagePreview:
                conversation.lastMessagePreview ?? existing.lastMessagePreview,
            lastMessageAt: conversation.lastMessageAt ?? existing.lastMessageAt,
          );
    return ConversationState(
        conversations: sortConversations([merged, ...others]));
  });
}

/// Tail update only (no re-insert for unknown conversations).
void touchConversation(
    Store<ConversationState> store, String key, String preview, int at) {
  store.update((prev) => ConversationState(
        conversations: sortConversations(prev.conversations.map((c) {
          if (c.key != key) return c;
          return c.copyWith(lastMessagePreview: preview, lastMessageAt: at);
        }).toList()),
      ));
}

void bumpUnread(Store<ConversationState> store, String key) {
  store.update((prev) => ConversationState(
        conversations: prev.conversations.map((c) {
          if (c.key != key) return c;
          return c.copyWith(unreadLocal: c.unreadLocal + 1);
        }).toList(),
      ));
}

void clearUnread(Store<ConversationState> store, String key) {
  store.update((prev) => ConversationState(
        conversations: prev.conversations.map((c) {
          if (c.key != key) return c;
          return c.copyWith(unreadLocal: 0);
        }).toList(),
      ));
}

/// Drop a conversation entirely (left the group, or kicked from it).
void removeConversation(Store<ConversationState> store, String key) {
  store.update((prev) {
    if (!prev.conversations.any((c) => c.key == key)) return prev;
    return ConversationState(
        conversations: prev.conversations.where((c) => c.key != key).toList());
  });
}

List<Conversation> sortConversations(List<Conversation> conversations) {
  final copy = List.of(conversations);
  copy.sort((a, b) {
    final byTime = (b.lastMessageAt ?? 0).compareTo(a.lastMessageAt ?? 0);
    return byTime != 0 ? byTime : a.key.compareTo(b.key);
  });
  return copy;
}

/// firstOrNull lives in package:collection; a private twin avoids the dep.
extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}
